umbrella.d.ts

```diff
@@ -1,207 +1,1484 @@
+     *     <li data-id='1'>Second</li>
+     *     <li data-id='2'>Third</li>
+     *   </ul>;
+     *
+     *   // Get the value for data-id:
+     *   u('ul li').first().data('id');
+     *   // 0
+     *
+     *   // Set the data-id of an element:
+     *   u('ul li').first().data({ id: '1' });
+     *   // <li data-id='1'>First</li>
+     *
+     *   u('ul li').first().data('id', '2');
+     *   // <li data-id='2'>First</li>
+     *
+     * @template NODE - Defines the type of DOM elements that the Umbrella instance manipulates.
+     * @type {this['attr']}
+     */
+    data: Umbrella<NODE>['attr'];
 
-    clone(): this;
+    /**
+     * Loop through all the nodes and execute a callback for each.
+     *
+     * [Documentation](https://umbrellajs.com/documentation#each)
+     *
+     * @example
+     *   // Loop through all the links and add them a target="_blank":
+     *   u('a').each(node => {
+     *     u(node).attr({ target: '_blank' });
+     *   });
+     *
+     * @template NODE - The current iteration DOM element.
+     * @param {IterateCallback<NODE>} callback - A function to execute on each element in the set. The callback receives
+     *   the current element and its index as arguments.
+     * @returns {this} - Returns the instance of Umbrella for method chaining.
+     */
+    each(callback: IterateCallback<NODE>): Umbrella<NODE>;
 
-    closest(filter?: any): this;
+    /**
+     * Removes all child nodes from each element in the matched set.
+     *
+     * [Documentation](https://umbrellajs.com/documentation#empty)
+     *
+     * @example
+     *   // Removes all child nodes from all containers:
+     *   u('div.container').empty();
+     *
+     * @template NODE - Defines the type of DOM elements that the Umbrella instance manipulates.
+     * @returns {this} - Returns the instance of Umbrella for method chaining.
+     */
+    empty(): Umbrella<NODE>;
+
+    /**
+     * Remove all the nodes that doesn't match the criteria.
+     *
+     * [Documentation](https://umbrellajs.com/documentation#filter)
+     *
+     * @example
+     *   // Get all the paragraphs with a link:
+     *   const paragraphs = u('p').filter(node => {
+     *     return u(node).find('a').length > 0;
+     *   });
+     *
+     *   // Get only the inputs with an answer above 5 and show an error:
+     *   u('input')
+     *     .filter(node => {
+     *       if (parseInt(u(node).first().value) > 5) {
+     *         return true;
+     *       }
+     *     })
+     *     .addClass('error');
+     *
+     * @template NODE - The current iteration DOM element.
+     * @param {IterateCallback<NODE, Maybe<boolean>>} predicate - A function that tests each element in the set. It
+     *   should return `true` to keep the element, or `false` (or a falsy value) to remove it.
+     * @returns {this} - A new Umbrella instance containing the elements that passed the test.
+     */
+    filter(predicate: IterateCallback<NODE, Maybe<boolean>>): Umbrella<NODE>;
```

Comment:
I found that a `filter` predicate does not take the current Umbrella instance as the context of the function ([see](https://github.com/franciscop/umbrella/blob/7988c20815ae966d3603cbc684a1580107f711e5/src/plugins/filter/filter.js#L25)), but I think it'd be better to fix it for general consistency.

Reply:
Umbrella is _supposed_ to be composable, meaning `u(x) == u(u(x))`. The "instance" is not supposed to be important (except for the list of nodes, which is passed there).